src/Microsoft.AspNet.OData.Shared/ODataAPIHandler.cs

```diff
@@ -64,129 +69,245 @@ public abstract class ODataAPIHandler<TStructuralType>: IODataAPIHandler where T
+                object getObject = null;
+
+                if (odataIdContainerHandler != null)
                 {
-                    if (this.TryGet(keys, out returnedObject, out error) == ODataAPIResponseStatus.Success)
+                    handlerForGet = odataIdContainerHandler;
+                    getResponse = (ODataAPIResponseStatus)handlerForGet.GetType().GetMethod(nameof(TryGet)).Invoke(handlerForGet, handlerParams);
+                    getObject = handlerParams[1];
+                }
+
+                if (getResponse == ODataAPIResponseStatus.Success)
+                {
+                    operation = DataModificationOperationKind.Link;
+
+                    CopyProperties(getObject, resource, navigationProperties);
+
+                    object[] addRelatedObjectParams = new object[] { getObject, null };
+
+                    ODataAPIResponseStatus addRelatedObjectResponse = (ODataAPIResponseStatus)odataApiHandler.GetType().GetMethod(nameof(TryAddRelatedObject)).Invoke(odataApiHandler, addRelatedObjectParams);
+
+                    if (addRelatedObjectResponse == ODataAPIResponseStatus.Failure)
                     {
-                        return returnedObject;
+                        HandleFailedOperation(resource, operation, addRelatedObjectParams[1].ToString());
+                        failedOperation = true;
                     }
-                    else
+                }
+                else
+                {
+                    ODataAPIResponseStatus createObjectResponse = (ODataAPIResponseStatus)odataApiHandler.GetType().GetMethod(nameof(TryCreate)).Invoke(odataApiHandler, handlerParams);
+
+                    if (createObjectResponse == ODataAPIResponseStatus.Failure)
                     {
-                        if (this.TryCreate(keys, out returnedObject, out error) == ODataAPIResponseStatus.Success)
-                        {
-                            return returnedObject;
-                        }
-                        else
-                        {
-                            return null;
-                        }
+                        HandleFailedOperation(resource, operation, handlerParams[2].ToString());
+                        failedOperation = true;
                     }
                 }
-                else
+            }
+            catch (Exception ex)
+            {
+                HandleFailedOperation(resource, operation, ex.Message);
+            }
+        }
+
+        private static void HandleFailedOperation(object originalObject, DataModificationOperationKind operation, string errorMessage)
+        {
+            Type type = originalObject.GetType();
+            PropertyInfo[] properties = type.GetProperties();
+            PropertyInfo oDataInstanceAnnotationContainerPropertyInfo = properties.FirstOrDefault(s => s.PropertyType == typeof(ODataInstanceAnnotationContainer) || s.PropertyType == typeof(IODataInstanceAnnotationContainer));
```

Comment:
This means people have to include `ODataInstanceAnnotationContainer` in their CLR objects?

Reply:
Yes.
We will document it. That's the most sensible approach at the moment